File: compras-service/app/utils/logger.py
"""
Sistema de logging profesional con formato legible
"""
import logging
import sys
from datetime import datetime
from pathlib import Path
from logging.handlers import TimedRotatingFileHandler
from app.config import settings


class CustomFormatter(logging.Formatter):
    """Formateador personalizado para logs legibles"""
    
    def format(self, record):
        # Formato: [2025-10-15 16:27:00] INFO [compras-service]: mensaje
        timestamp = datetime.fromtimestamp(record.created).strftime('%Y-%m-%d %H:%M:%S')
        level = record.levelname
        
        # Construir mensaje base
        log_msg = f"[{timestamp}] {level:5s} [compras-service]: {record.getMessage()}"
        
        # Agregar información extra si existe
        if hasattr(record, 'request_id'):
            log_msg = f"[{timestamp}] {level:5s} [compras-service] [req:{record.request_id}]: {record.getMessage()}"
        
        # Agregar excepción si existe
        if record.exc_info:
            log_msg += "\n" + self.formatException(record.exc_info)
        
        return log_msg


def setup_logger():
    """Configura el sistema de logging profesional"""
    
    # Crear directorio de logs si no existe
    log_dir = Path("logs")
    log_dir.mkdir(exist_ok=True)
    
    # Configurar logger raíz
    logger = logging.getLogger()
    logger.setLevel(getattr(logging, settings.log_level.upper()))
    
    # Limpiar handlers existentes
    logger.handlers.clear()
    
    # ========== HANDLER PARA CONSOLA (con colores) ==========
    console_handler = logging.StreamHandler(sys.stdout)
    console_handler.setLevel(logging.DEBUG if settings.node_env == "development" else logging.INFO)
    console_formatter = CustomFormatter()
    console_handler.setFormatter(console_formatter)
    logger.addHandler(console_handler)
    
    # ========== HANDLER PARA ARCHIVO (rotación diaria) ==========
    file_handler = TimedRotatingFileHandler(
        filename=log_dir / "compras-service.log",
        when="midnight",
        interval=1,
        backupCount=14,  # Mantener 14 días
        encoding="utf-8"
    )
    file_handler.setLevel(logging.INFO)
    file_formatter = CustomFormatter()
    file_handler.setFormatter(file_formatter)
    logger.addHandler(file_handler)
    
    # ========== HANDLER PARA ERRORES (archivo separado) ==========
    error_handler = TimedRotatingFileHandler(
        filename=log_dir / "compras-service-errors.log",
        when="midnight",
        interval=1,
        backupCount=30,  # Mantener errores por 30 días
        encoding="utf-8"
    )
    error_handler.setLevel(logging.ERROR)
    error_formatter = CustomFormatter()
    error_handler.setFormatter(error_formatter)
    logger.addHandler(error_handler)
    
    # Silenciar logs muy verbosos de librerías
    logging.getLogger("uvicorn.access").setLevel(logging.WARNING)
    logging.getLogger("databases").setLevel(logging.WARNING)
    logging.getLogger("httpx").setLevel(logging.WARNING)
    logging.getLogger("httpcore").setLevel(logging.WARNING)
    
    logger.info("[compras-service]: Logger inicializado correctamente")
    
    return logger


# Instancia global del logger
logger = setup_logger()


# Funciones helper para logging estructurado
def log_request(method: str, path: str, status_code: int, duration_ms: float, user_id: str = None):
    """Registra una petición HTTP"""
    user_info = f" | Usuario: {user_id}" if user_id else ""
    logger.info(f"{method} {path} - {status_code} ({duration_ms:.2f}ms){user_info}")


def log_db_operation(operation: str, table: str, success: bool, duration_ms: float = None):
    """Registra operaciones de base de datos"""
    status = "exitosa" if success else "fallida"
    duration_info = f" ({duration_ms:.2f}ms)" if duration_ms else ""
    logger.info(f"Operación BD: {operation} en tabla '{table}' - {status}{duration_info}")


def log_external_call(service: str, endpoint: str, status_code: int = None, duration_ms: float = None):
    """Registra llamadas a servicios externos"""
    status_info = f" - {status_code}" if status_code else ""
    duration_info = f" ({duration_ms:.2f}ms)" if duration_ms else ""
    logger.info(f"Llamada externa a {service}: {endpoint}{status_info}{duration_info}")


def log_business_event(event: str, details: dict = None):
    """Registra eventos de negocio importantes"""
    details_str = f" | Detalles: {details}" if details else ""
    logger.info(f"Evento de negocio: {event}{details_str}")

File: servicio-notificaciones/src/utils/logger.ts
import winston from 'winston';
import DailyRotateFile from 'winston-daily-rotate-file';
import config from '../config/config';
import path from 'path';
import fs from 'fs';

// Crear directorio de logs si no existe
const logsDir = path.join(__dirname, '../../logs');
if (!fs.existsSync(logsDir)) {
    fs.mkdirSync(logsDir, { recursive: true });
}

// Formato personalizado para logs legibles
const customFormat = winston.format.combine(
    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
    winston.format.errors({ stack: true }),
    winston.format.printf(({ timestamp, level, message, stack, ...metadata }) => {
        let log = `[${timestamp}] ${level.toUpperCase().padEnd(5)} [notificaciones-service]: ${message}`;
        
        // Agregar metadata si existe (excepto timestamp y level)
        const metaKeys = Object.keys(metadata).filter(key => key !== 'timestamp' && key !== 'level');
        if (metaKeys.length > 0) {
            const metaStr = metaKeys
                .map(key => `${key}=${JSON.stringify(metadata[key])}`)
                .join(', ');
            log += ` | ${metaStr}`;
        }
        
        // Agregar stack trace si existe
        if (stack) {
            log += `\n${stack}`;
        }
        
        return log;
    })
);

// Formato para consola con colores
const consoleFormat = winston.format.combine(
    winston.format.colorize(),
    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
    winston.format.printf(({ timestamp, level, message }) => {
        return `[${timestamp}] ${level}: ${message}`;
    })
);

// ========== TRANSPORTS ==========

// Consola (con colores)
const consoleTransport = new winston.transports.Console({
    format: consoleFormat,
    level: config.nodeEnv === 'development' ? 'debug' : 'info',
});

// Archivo general (rotación diaria)
const fileTransport: DailyRotateFile = new DailyRotateFile({
    filename: path.join(logsDir, 'notificaciones-service-%DATE%.log'),
    datePattern: 'YYYY-MM-DD',
    maxSize: '20m',
    maxFiles: '14d', // Mantener 14 días
    format: customFormat,
    level: 'info',
});

// Archivo de errores (rotación diaria)
const errorFileTransport: DailyRotateFile = new DailyRotateFile({
    filename: path.join(logsDir, 'notificaciones-service-errors-%DATE%.log'),
    datePattern: 'YYYY-MM-DD',
    maxSize: '20m',
    maxFiles: '30d', // Mantener errores por 30 días
    format: customFormat,
    level: 'error',
});

// ========== CREAR LOGGER ==========
const logger = winston.createLogger({
    level: config.logging.level,
    transports: [
        consoleTransport,
        fileTransport,
        errorFileTransport,
    ],
    exitOnError: false,
});

// ========== FUNCIONES HELPER ==========

/**
 * Log de mensaje recibido desde RabbitMQ
 */
export function logMensajeRecibido(compraId: number, usuarioId: string, eventoId: number): void {
    logger.info(`Mensaje recibido de RabbitMQ | CompraID: ${compraId} | Usuario: ${usuarioId} | Evento: ${eventoId}`);
}

/**
 * Log de llamada a servicio externo
 */
export function logExternalCall(service: string, endpoint: string, statusCode?: number, durationMs?: number): void {
    const status = statusCode ? ` - ${statusCode}` : '';
    const duration = durationMs ? ` (${durationMs.toFixed(2)}ms)` : '';
    logger.info(`Llamada externa a ${service}: ${endpoint}${status}${duration}`);
}

/**
 * Log de evento de negocio
 */
export function logBusinessEvent(event: string, details: Record<string, any>): void {
    const detailsStr = Object.entries(details)
        .map(([key, value]) => `${key}="${value}"`)
        .join(', ');
    logger.info(`Evento de negocio: ${event} | ${detailsStr}`);
}

/**
 * Log de email enviado
 */
export function logEmailEnviado(email: string, compraId: number, messageId: string): void {
    logger.info(`Email enviado exitosamente | Destinatario: ${email} | CompraID: ${compraId} | MessageID: ${messageId}`);
}

/**
 * Log de error de email
 */
export function logEmailError(email: string, compraId: number, error: string): void {
    logger.error(`Error al enviar email | Destinatario: ${email} | CompraID: ${compraId} | Error: ${error}`);
}

/**
 * Log de conexión RabbitMQ
 */
export function logRabbitMQConexion(queueName: string, success: boolean): void {
    if (success) {
        logger.info(`Conectado a RabbitMQ exitosamente | Cola: ${queueName}`);
    } else {
        logger.error(`Error al conectar a RabbitMQ | Cola: ${queueName}`);
    }
}

/**
 * Log de inicio del servicio
 */
export function logServiceStartup(): void {
    logger.info('='.repeat(60));
    logger.info('[notificaciones-service]: Iniciando Servicio de Notificaciones');
    logger.info(`[notificaciones-service]: Ambiente: ${config.nodeEnv}`);
    logger.info(`[notificaciones-service]: Nivel de logs: ${config.logging.level}`);
    logger.info('='.repeat(60));
}

/**
 * Log de cierre del servicio
 */
export function logServiceShutdown(): void {
    logger.info('[notificaciones-service]: Servicio de notificaciones cerrado correctamente');
}

export default logger;